api/src/main/java/org/eclipse/microprofile/metrics/DefaultMetadata.java

```diff
@@ -0,0 +1,178 @@
+     * </p>
+     */
+    private final MetricType type;
+    /**
+     * Unit of the metric.
+     * <p>
+     * An optional field which holds the Unit of the metric object.
+     * </p>
+     */
+    private final String unit;
+
+    /**
+     * Can this metric name (in a scope) be used multiple times?
+     * <p>
+     * Setting this is optional. The default is <tt>false</tt>, which
+     * prevents reusing.
+     * </p>
+     * Note that this only has an effect if the <tt>name</tt> is explicitly given or
+     * <tt>absolute</tt> is set to true and two methods that are marked as metric have
+     * the same name.
+     * <p>
+     * If the name is automatically determined, then this flag has no effect as
+     * all metric names are different anyway
+     */
+    private final boolean reusable;
+
+    /**
+     * Tags of the metric. Augmented by global tags.
+     * <p>
+     * An optional field which holds the tags of the metric object which can be
+     * augmented by global tags.
+     * </p>
+     */
+    private final Map<String, String> tags;
+
+    public DefaultMetadata(String name, String displayName, String description,
+                           MetricType type, String unit, boolean reusable,
+                           Map<String, String> tags) {
+        this.name = name;
+        this.displayName = displayName;
+        this.description = description;
+        this.type = type;
+        this.unit = unit;
+        this.reusable = reusable;
+        this.tags = tags;
+    }
+
+    @Override
+    public String getName() {
+        return name;
+    }
+
+    @Override
+    public String getDisplayName() {
+        return Optional.ofNullable(displayName).orElse(name);
+    }
+
+    @Override
+    public Optional<String> getDescription() {
+        return Optional.ofNullable(description);
+    }
+
+    @Override
+    public String getType() {
+        return Optional.ofNullable(type).orElse(MetricType.INVALID).toString();
+    }
+
+    @Override
+    public MetricType getTypeRaw() {
+        return Optional.ofNullable(type).orElse(MetricType.INVALID);
+    }
+
+    @Override
+    public Optional<String> getUnit() {
+        return Optional.ofNullable(unit);
+    }
+
+    @Override
+    public boolean isReusable() {
+        return reusable;
+    }
+
+    @Override
+    public String getTagsAsString() {
+        return this.tags.entrySet()
+                .stream().map(e -> e.getKey() + "=\"" + e.getValue() + "\"")
+                .collect(Collectors.joining(","));
+    }
+
+    @Override
+    public Map<String, String> getTags() {
+        return Collections.unmodifiableMap(tags);
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) {
+            return true;
+        }
+        if (!(o instanceof Metadata)) {
+            return false;
+        }
+        Metadata that = (Metadata) o;
+        return Objects.equals(name, that.getName())
+                && type == that.getTypeRaw();
```

Comment:
Per @pilhuhn 's comment no two metrics may have the same name but different types.  I think that should mean that two metrics with the same name are the same metric -- and that type shouldn't be used to make that determination.